### PR TITLE
Add global config option instances.nic.host_name

### LINKED ISSUE #10036: Add global config option instances.network.host_name to control NIC host_name generation
Add an `instances.network.host_name` global config option that can have `random` or `mac`, with a default value of `random`.

If it is set to `random` then use the random host interface names as we do today, but if it's set to mac, then generate a name in the form `lxd<mac_addres>`, e.g. `lxd001122334455` which just happens to fit the 15 bytes we have.

### PR DESCRIPTION
**Pull request by presztak**
Closes #10036

### CODE REVIEW

**Comment by stgraber** [comment]
This PR adds the new config key as an instance option whereas the issue asks for it to be added as a global config option (server config).

The reason for wanting it to be a global option is that we don't really want a mix of naming schemes on the system nor do we want to allow restricted users to choose between the two schemes.

**Comment by presztak** [comment]
Oh, I misunderstood. I will fix this soon.

**Comment by stgraber** [comment]
No worries, thanks!

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
@tomponline can you take a quick look at this?

**Comment by presztak** [comment]
There was a typo in `doc/server.md`. I fixed this.

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/device/device_common.go [hunk 1]
```diff
@@ -94,3 +97,13 @@ func (d *deviceCommon) Update(oldDevices deviceConfig.Devices, isRunning bool) e
 func (d *deviceCommon) Remove() error {
 	return nil
 }
+
+// hostNameMode returns host name mode from cluster config.
+func (d *deviceCommon) hostNameMode() (string, error) {
```
[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/device/device_common.go:102]: You can then replace calls to `network.RandomDevName()` with `d.generateHostName()`.

File: doc/server.md [hunk 2]
```diff
@@ -52,6 +53,7 @@ images.auto\_update\_interval       | integer   | global    | 6
 images.compression\_algorithm       | string    | global    | gzip                              | Compression algorithm to use for new images (bzip2, gzip, lzma, xz or none)
 images.default\_architecture        | string    | -         | -                                 | Default architecture which should be used in mixed architecture cluster
 images.remote\_cache\_expiry        | integer   | global    | 10                                | Number of days after which an unused cached remote image will be flushed
+instances.network.host\_name        | string    | global    | random                            | If it is set to `random` then use the random host interface names but if it's set to mac, then generate a name in the form `lxd<mac_address>`(MAC without leading 2 digits).
```
Comment (tomponline) [inline doc/server.md:56]: @stgraber do you think we should use `instanes.nic.host_name` to indicate it controls the NIC host_name generation? Using "network" here seems a bit confusing.

Comment (stgraber) [reply doc/server.md:56]: Sure, I'm fine with using `nic`

Comment (presztak) [reply doc/server.md]: Fixed.

File: lxd/device/device_common.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (presztak) [reply lxd/device/device_common.go]: Fixed.

File: lxd/network/network_utils.go [hunk 3]
```diff
@@ -49,10 +49,18 @@ func networkValidPort(value string) error {
 	return nil
 }
 
-// RandomDevName returns a random device name with prefix.
+// RandomDevName returns a random device name with prefix if mode is equal 'random'.
+// For mode equal 'mac' returns interface name with prefix 'lxd' and MAC without leading 2 digits.
 // If the random string combined with the prefix exceeds 13 characters then empty string is returned.
 // This is to ensure we support buggy dhclient applications: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=858580
-func RandomDevName(prefix string) string {
+func RandomDevName(prefix string, mode string, hwaddr string) string {
```
[the remark this answers is not in the document]
Comment (presztak) [reply lxd/network/network_utils.go]: Fixed.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Thanks!

**Review by tomponline — COMMENTED** [review]
LGTM me, thanks!

Its missing any tests though so I'll add some.

**Pull request merged**